### ISSUE TITLE
Spotify is broken

### ISSUE THREAD

**Issue by rvolosatovs** [NixOS/nixpkgs#39208]
## Issue description
Spotify in 18.03 channel fails with:
```log
$ spotify
/home/rvolosatovs/.nix-profile/bin/spotify: /nix/store/phvca4sgam9a5lf5z4wviqwvlcwq6lk8-curl-7.59.0/lib/libcurl-gnutls.so.4: no version information available (required by /home/rvolosatovs/.nix-profile/bin/spotify)
/nix/store/4535qsyqvx141wvch8626i6m5qibkim8-spotify-1.0.77.338.g758ebd78-41/share/spotify/.spotify-wrapped: /nix/store/phvca4sgam9a5lf5z4wviqwvlcwq6lk8-curl-7.59.0/lib/libcurl-gnutls.so.4: no version information available (required by /nix/store/4535qsyqvx141wvch8626i6m5qibkim8-spotify-1.0.77.338.g758ebd78-41/share/spotify/.spotify-wrapped)
/proc/self/exe: /nix/store/phvca4sgam9a5lf5z4wviqwvlcwq6lk8-curl-7.59.0/lib/libcurl-gnutls.so.4: no version information available (required by /proc/self/exe)
Fontconfig warning: "/etc/fonts/fonts.conf", line 146: blank doesn't take any effect anymore. please remove it from your fonts.conf
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  set_option: Cannot assign requested address
[1]    10545            spotify
```

Seems to be caused by https://github.com/NixOS/nixpkgs/pull/38874

### Steps to reproduce
Install Spotify from `channels/release-18.03`.
`$ spotify`

## Technical details
```log
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.34, NixOS, 18.03.132021.c0c5571ec1a (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0`
 - channels(root): `"nixos-18.03, nixpkgs-18.09pre137243.31effc1d7d5"`
 - channels(rvolosatovs): `""`
 - nixpkgs: `/nix/nixpkgs`
```

**Comment by joachifm** [comment]
@utdemir & @srhb might know more about this

**Comment by srhb** [comment]
I believe the gnutls warning is harmless and has been there before.

However, I can not reproduce the boost error, and spotify launches for me just fine. Hmm!

**Comment by srhb** [comment]
@rvolosatovs Long shot, do you have any tun network devices?

**Comment by srhb** [comment]
@rvolosatovs And can you try to launch spotify with --force-unicast-discovery ?

**Comment by ajs124** [comment]
I'm hitting this error on one of my systems but not on another. Both are on 4535qsyqvx141wvch8626i6m5qibkim8-spotify-1.0.77.338.g758ebd78-41.

They both share a lot of config, so I'm not sure how and why they would be different for this to be happening on one of them, but not the other.

**Comment by ajs124** [comment]
Already did, see updated comment.

I'm running wireguard, but on both systems. The affected system also has an ethernet, wifi and umts network interface.

**Comment by srhb** [comment]
@ajs124 Thank you!

**Comment by utdemir** [comment]
Sorry, I am the author of that update, but I don't have any insights what is going on. 

I just updated the version, it complained about not finding `libcurl-gnutls.so.4`, and I overrode `curl` to have `gnutls` support. I don't think reverting the patch is an option since Spotify already deleted the old version from their repository.

**Comment by ajs124** [comment]
@rvolosatovs findings seem reproducible on my setup. Spotify runs with Wifi *and* Ethernet up. What?

**Comment by xylo** [comment]
I'm not using NixOS but I discovered the same bug with the spotify client since my last update:

> what():  set_option: Cannot assign requested address

So this bug has definitely nothing to do with curl and gnutls.  It's a bug in the latest deb version of the spotify client and has already been reported in the spotify forum: https://community.spotify.com/t5/Desktop-Linux/Spotify-apt-hangs-on-startup-in-Ubuntu-18-04/m-p/4461877

**Comment by srhb** [comment]
I think that we've somehow ended up on the testing deb again despite the warning about it, but it appears stable is broken in other ways. Could be worth investigating though.

**Comment by ajs124** [comment]
So, someone in one of those came up with [an LD_PRELOAD hack](https://gist.github.com/mdierolf/7751f9c4535a6d402438e219617cf4c3), which seems to work for me.

Overriding setsockopt isn't the nicest thing, but then, what about this application is?

**Comment by ajs124** [comment]
It's fixed in 1.0.79.223.g92622cc2-21

**Comment by milahu** [comment]
> it complained about not finding `libcurl-gnutls.so.4`, and I overrode `curl` to have `gnutls` support

sounds like the obvious solution (?)

i ran into the same problem trying to wrap spotify with `LD_PRELOAD=spotify-adblock.so`

```
LD_LIBRARY_PATH='... all the paths generated by `( lib.makeLibraryPath deps )` ...' \
ldd /nix/store/*-spotify-unwrapped-*/share/spotify/.spotify-wrapped | grep not

 libcurl-gnutls.so.4 => not found
```

solved like [nixpkgs/pkgs/applications/video/filebot/default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/filebot/default.nix)

```
  curlWithGnutls = curl.override { gnutlsSupport = true; sslSupport = false; };
```
